### PR TITLE
Add ability to configure git tool, repository browser, and additional behaviours for literate/multi-branch projects.

### PR DESCRIPTION
**Pull request by wadahiro**
This can be useful for some jobs for multi-branch projects (such as literate plugin and multi-branch project plugin).

### CODE REVIEW

**Comment by MarkEWaite** [comment]
@jglick and @ndeloof I need your help to evaluate this pull request.  I'm not a credible reviewer of a proposal to extend the git plugin to assist with workflow.

**Comment by ndeloof** [comment]
@wadahiro can you prepare a PR on literate/multi-branch to demonstrate use of this ?
It's not clear to me how this is supposed to be used. Not opposed to a merge, but seems unnecessary from first overview

**Comment by wadahiro** [comment]
@ndeloof I changed this PR description. How about this?

**Comment by ndeloof** [comment]
@wadahiro the description is fine, this is not what I was asking for. Changes looks good, but _why_ they are needed to is not clear to me, so a sample usage from another plugin would help.

**Comment by wadahiro** [comment]
@ndeloof 
I am sorry for not explaining sufficiently.

> why they are needed to is not clear to me, 

Some plugins configure git configurations via scm-api (the implementation is GitSCMSource). As far as I know, [Literate plugin](https://github.com/jenkinsci/literate-plugin) and [Multi-branch project plugin](https://github.com/mjdetullio/multi-branch-project-plugin) use this scm-api. Because of GitSCMSource doesn't support more detailed settings such as 'Additional Behaviors' currently, we cannot use useful options such as 'Shallow clone'.

> so a sample usage from another plugin would help.

In [Multi-branch project plugin](https://github.com/mjdetullio/multi-branch-project-plugin), it includes 'descriptor.configPage' of SCMSource simply [in the config jelly](https://github.com/mjdetullio/multi-branch-project-plugin/blob/master/src/main/resources/com/github/mjdetullio/jenkins/plugins/multibranch/AbstractMultiBranchProject/configure-entries.jelly#L74-L75). So if GitSCMSource supports more options, it can be generate more options automatically.

**Comment by ndeloof** [comment]
got it

LGTM

**Comment by wadahiro** [comment]
@ndeloof When will this PR be merged? Has there been any issue with this PR itself? Please let me know if there are any problems.

**Comment by ndeloof** [comment]
I'm not maintaining this plugin anymore

**Comment by wadahiro** [comment]
@MarkEWaite Could you evaluate this pull request? Please let me know if there are any problems.

File: src/main/java/jenkins/plugins/git/GitSCMSource.java [hunk 1]
```diff
@@ -75,14 +82,25 @@
...
+    private final GitRepositoryBrowser browser;
+
+    private final String gitTool;
+
+    private final List<GitSCMExtension> extensions;
+
     @DataBoundConstructor
-    public GitSCMSource(String id, String remote, String credentialsId, String includes, String excludes, boolean ignoreOnPushNotifications) {
+    public GitSCMSource(String id, String remote, String credentialsId, String includes, String excludes,
+            boolean ignoreOnPushNotifications, GitRepositoryBrowser browser, String gitTool,
```
Comment (KostyaSha) [inline src/main/java/jenkins/plugins/git/GitSCMSource.java]: Use `@DataBoundSetters` ?

Comment (wadahiro) [inline src/main/java/jenkins/plugins/git/GitSCMSource.java]: @KostyaSha Thanks. I'll fix to use @DataBoundSetters against two new args(browser, gitTool).

**Comment by MarkEWaite** [comment]
I hope to evaluate this pull request (code review and interactive test) within the next week. Thanks for your patience.

**Comment by jason-womack** [comment]
@MarkEWaite What is the status of this PR?

**Comment by MarkEWaite** [comment]
It will likely be many weeks before this pull request is evaluated. My optimism in October has been tempered by a very heavy schedule at work.

**Comment by bgianfo** [comment]
@wadahiro Any hope of syncing this up to master?

**Comment by MarkEWaite** [comment]
Yes, there's hope. The merge conflict is trivial enough that I can resolve it myself.  I doubt it will be on the master today, but it may be on the master within the next 1-2 weeks.

**Comment by bgianfo** [comment]
Anything I can do to help?

**Comment by MarkEWaite** [comment]
Sure @bgianfo , you can help by building a copy of this pull request and deploying it.  Test it in various ways to confirm that it does not break existing use models, then post your results with the pull request. I'm much more comfortable when others can say that they've used a pull request successfully.

**Comment by wadahiro** [comment]
@MarkEWaite I resolved the merge conflict.
I have been using this patched version with [Multi-branch project plugin](https://github.com/mjdetullio/multi-branch-project-plugin) for six months in the several projects, but there was no problem.

**Comment by MarkEWaite** [comment]
@wadahiiro, thanks for resolving the merge conflicts. I'm glad to hear that it has been working well for you.  It will be several weeks (at least) before I can evaluate this.  

I'm traveling on business this week and won't have much time for Jenkins work while I'm traveling.  

I will start the automated evaluation in my environment later today so that I can at least report if the automated tests were successful in the multi-platform / multi-version automated test environment.

**Comment by wadahiro** [comment]
@MarkEWaite Any updates?

**Comment by wadahiro** [comment]
@MarkEWaite Thanks!! 😄

**Comment by oleg-nenashev** [comment]
This change breaks the binary compatibility and causes issues like [JENKINS-36419](https://issues.jenkins-ci.org/browse/JENKINS-36419). All abstract methods need to be reworked

**Comment by oleg-nenashev** [comment]
Working on the fix, but such interface is going to be a headache. Seems I'll have to just move some methods upstairs and break the compatibility for setters

**Comment by oleg-nenashev** [comment]
@wadahiro Do you use this functionality anywhere?

File: src/main/java/jenkins/plugins/git/GitSCMSource.java [hunk 2]
```diff
@@ -91,6 +109,42 @@ public boolean isIgnoreOnPushNotifications() {
...
+    public void setGitTool(String gitTool) {
+        this.gitTool = gitTool;
+    }
+
+    @Override
+    public List<GitSCMExtension> getExtensions() {
+        if (extensions == null) {
+            extensions = new ArrayList<GitSCMExtension>();
+        }
+        return extensions;
```
Comment (oleg-nenashev) [inline src/main/java/jenkins/plugins/git/GitSCMSource.java]: This method exposes internal implementation. FindBugs should not have accepted it. CC @MarkEWaite  (it it still disabled?)

Comment (MarkEWaite) [inline src/main/java/jenkins/plugins/git/GitSCMSource.java]: Yes, it is still disabled.  I'll check it more rigorously in the future. Sorry about that!

**Comment by wadahiro** [comment]
@oleg-nenashev @MarkEWaite I apologize for the breaking change. :bow:  I shouldn't have add abstract methods into AbstractGitSCMSource class. Thank you for your great works!

**Pull request merged**